### PR TITLE
Update EXPECTED_XCVR_THRESHOLD_INFO_KEYS list in test_sfp.py

### PR DESCRIPTION
**Pull request by rajann**
### Description of PR

Updated  EXPECTED_XCVR_THRESHOLD_INFO_KEYS list with additional fields from QSFP-DD modules

### Back port request
- [ ] 201911
- [ ] 202012
- [x] 202205

### CODE REVIEW

**Comment by prgeor** [comment]
@rajann can you run the same test for QSFP28 or QSFP+

**Comment by mannytaheri** [comment]
Even with QSFP-DD the fields are different for different types of SFPs:  Please check the fields for different types of SFPs.

"""
EXPECTED_XCVR_THRESHOLD_INFO_KEYS for SR8 SFP
'txpowerlowwarning',
'temphighwarning',
'temphighalarm',
'txbiashighalarm',
'vcchighalarm',
'txbiaslowalarm',
'rxpowerhighwarning',
'vcclowwarning',
'txbiaslowwarning',
'rxpowerlowalarm',
'vcchighwarning',
'txpowerhighwarning',
'rxpowerlowwarning',
'txbiashighwarning',
'vcclowalarm',
'txpowerhighalarm',
'templowalarm',
'rxpowerhighalarm',
'templowwarning',
'txpowerlowalarm'
"""

"""
EXPECTED_XCVR_THRESHOLD_INFO_KEYS for ZR SFP
'lasertemplowwarning'
'rxsigpowerhighalarm' 
'osnrhighwarning'
'vcclowwarning'
'txbiaslowwarning'
'sopmdlowalarm'
'biasyqlowwarning'
'osnrlowwarning'
 'pdllowalarm'
'dgdlowalarm' 
'biasxihighwarning'
'osnrhighalarm'
'txpowerlowalarm'
'templowalarm'
'biasyplowalarm' 
'prefecberhighwarning'
'postfecberlowwarning'
'dgdlowwarning'
'txpowerhighwarning'
'cfolowalarm'
'postfecberlowalarm'
'txpowerhighalarm'
'rxtotpowerhighwarning'
'temphighwarning'
'temphighalarm'
'cdlonglowalarm'
'cdshortlowalarm'
'postfecberhighwarning'
'rxtotpowerlowwarning'
'biasyilowwarning'
'pdlhighwarning'
'biasxqhighwarning'
'rxpowerhighalarm'
'txcurrpowerlowwarning'
'osnrlowalarm' 
'rxpowerlowalarm'
'biasxqlowalarm'
'txcurrpowerhighalarm'
'lasertemplowalarm'
'sopmdhighalarm'
'biasyplowwarning'
'biasyqhighalarm'
'cdlonghighalarm'
'vcchighalarm'
'txcurrpowerhighwarning'
'pdlhighalarm'
'rxtotpowerlowalarm'
'cdshortlowwarning'
'biasxplowalarm'
'txbiashighalarm'
'biasyphighalarm'
'biasyihighwarning'
'cfohighwarning'
'cfohighalarm'
'txbiaslowalarm'
'dgdhighwarning'
'vcclowalarm'
'cdlonghighwarning'
'biasxihighalarm'
'esnrlowwarning'
'esnrlowalarm'
'biasxplowwarning'
'prefecberhighalarm'
'txbiashighwarning'
'biasxilowalarm'
'rxtotpowerhighalarm'
'pdllowwarning'
'postfecberhighalarm'
'biasxphighwarning'
'txcurrpowerlowalarm'
'biasyqlowalarm'
'biasyphighwarning'
'biasxphighalarm'
'rxsigpowerhighwarning'
'esnrhighalarm'
'cdshorthighwarning'
'cdlonglowwarning'
'biasxqhighalarm'
'prefecberlowalarm'
'txpowerlowwarning'
'esnrhighwarning'
'sopmdhighwarning'
'cdshorthighalarm'
'biasxilowwarning'
'sopmdlowwarning'
'prefecberlowwarning'
'lasertemphighwarning'
'lasertemphighalarm'
'vcchighwarning'
'biasyqhighwarning'
'templowwarning'
'rxpowerhighwarning'
'biasyilowalarm'
'biasyihighalarm'
'biasxqlowwarning'
'rxpowerlowwarning'
'rxsigpowerlowwarning'
'rxsigpowerlowalarm'
'dgdhighalarm'
'cfolowwarning'
"""

**Comment by rajann** [comment]
Should we add all the "Transceiver DOM Threshold Table" from https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#213-transceiver-dom-threshold-table into this expected_threshold_info_key

**Comment by prgeor** [comment]
> Should we add all the "Transceiver DOM Threshold Table" from https://github.com/sonic-net/SONiC/blob/master/doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md#213-transceiver-dom-threshold-table into this expected_threshold_info_key

The same test should pass for QSFP28, QSFP-DD and ZR.. May be better to have different tables for QSFP28, QSFP-DD and ZR. if so can update the test?

**Comment by rajann** [comment]
Verified the patch with both QSFP28 and QSFP-DD modules. I don't have access to ZR optics, hence i couldn't verify it.

File: tests/platform_tests/api/test_sfp.py [hunk 1]
```diff
@@ -322,16 +452,30 @@ def test_get_transceiver_threshold_info(self, duthosts, enum_rand_one_per_hwsku_
...
+                continue
+
             thold_info_dict = sfp.get_transceiver_threshold_info(platform_api_conn, i)
             if self.expect(thold_info_dict is not None, "Unable to retrieve transceiver {} threshold info".format(i)):
                 if self.expect(isinstance(thold_info_dict, dict), "Transceiver {} threshold info appears incorrect".format(i)):
                     actual_keys = thold_info_dict.keys()
 
-                    missing_keys = set(self.EXPECTED_XCVR_THRESHOLD_INFO_KEYS) - set(actual_keys)
+                    if info_dict["type_abbrv_name"] == "QSFP-DD":
+                        if 'ZR' in info_dict[media_interface_code]:
```
Comment (mannytaheri) [inline tests/platform_tests/api/test_sfp.py:467]: media_interface_code needs to be in a quote:   info_dict['media_interface_code']

File: tests/platform_tests/api/test_sfp.py [hunk 2]
```diff
@@ -151,6 +151,136 @@ class TestSfpApi(PlatformApiTestBase):
...
+        'rxtotpowerhighalarm',
+        'rxtotpowerlowalarm',
+        'rxtotpowerhighwarning',
+        'rxtotpowerlowwarning',
+        'rxsigpowerhighalarm',
+        'rxsigpowerlowalarm',
+        'rxsigpowerhighwarning',
+        'rxsigpowerlowwarning'
+    ]
+
```
Comment (mannytaheri) [inline tests/platform_tests/api/test_sfp.py:283]: Seems like  QSFPDD table  = XCVR table + new fields

and 

QSFPZR table =  QSFPDD table + new fields.

Might make sense to have only the new fields in QSPFDD and QSPFZR tables.

So, code below might look like:

```
expected_keys = self.EXPECTED_XCVR_THRESHOLD_INFO_KEYS
if info_dict["type_abbrv_name"] == "QSFP-DD":
  expected_keys += self.QSFPDD_EXPECTED_XCVR_THRESHOLD_INFO_KEYS
  if 'ZR' in info_dict["media_interface_code"]:
     expected_keys += self.QSFPZR_EXPECTED_XCVR_THRESHOLD_INFO_KEYS
```

Comment (prgeor) [reply tests/platform_tests/api/test_sfp.py:283]: agree with @mannytaheri. @rajann can you make this change?

Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:283]: Sure..I can make this change.  Hoping this list always backward compatible.

File: tests/platform_tests/api/test_sfp.py [hunk 3]
```diff
@@ -322,16 +414,28 @@ def test_get_transceiver_threshold_info(self, duthosts, enum_rand_one_per_hwsku_
...
+                logger.info("test_get_transceiver_threshold_info: Skipping transceiver {} (not applicable for this transceiver type)".format(i))
+                continue
+
             thold_info_dict = sfp.get_transceiver_threshold_info(platform_api_conn, i)
             if self.expect(thold_info_dict is not None, "Unable to retrieve transceiver {} threshold info".format(i)):
                 if self.expect(isinstance(thold_info_dict, dict), "Transceiver {} threshold info appears incorrect".format(i)):
                     actual_keys = thold_info_dict.keys()
 
-                    missing_keys = set(self.EXPECTED_XCVR_THRESHOLD_INFO_KEYS) - set(actual_keys)
+                    expected_keys = self.EXPECTED_XCVR_THRESHOLD_INFO_KEYS
```
[the remark this answers is not in the document]
Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:428]: I don't think so. As per the flow we are re-initializing the list of expected keys for every port from the for loop right?
Transceiver info is retrieved for every optics and this list is re-initialized

[the remark this answers is not in the document]
Comment (mihirpat1) [reply tests/platform_tests/api/test_sfp.py:428]: > I don't think so. As per the flow we are re-initializing the list of expected keys for every port from the for loop right? Transceiver info is retrieved for every optics and this list is re-initialized

As part of the for loop, we are just pointing expected_keys to EXPECTED_XCVR_THRESHOLD_INFO_KEYS list (**pass by reference and not by value**) every time. Hence, any corresponding list related changes using expected_keys will directly change the EXPECTED_XCVR_THRESHOLD_INFO_KEYS list (this currently will be done through "expected_keys += self.QSFPDD_EXPECTED_XCVR_THRESHOLD_INFO_KEYS" or "expected_keys += self.QSFPZR_EXPECTED_XCVR_THRESHOLD_INFO_KEYS").

So if you try the test I mentioned earlier, you would now see that for the non QSFP-DD,  EXPECTED_XCVR_THRESHOLD_INFO_KEYS will by default include fields for QSFPDD and/or QSFPZR even though, we would not execute line 430 and 432 for that iteration.

Comment (mihirpat1) [inline tests/platform_tests/api/test_sfp.py:428]: Also, please rename EXPECTED_XCVR_THRESHOLD_INFO_KEYS to something like EXPECTED_XCVR_COMMON_THRESHOLD_INFO_KEYS or so.

[the remark this answers is not in the document]
Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:428]: yeah that's right. It's pass by reference and the list keep growing. Modified to re-init the expected_keys at every iteration. And verified the logs for the list during the iteration.

Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:428]: Done

**Review by prgeor — DISMISSED** [review]

File: tests/platform_tests/api/test_sfp.py [hunk 4]
```diff
@@ -262,7 +354,6 @@ def test_get_transceiver_info(self, duthosts, enum_rand_one_per_hwsku_hostname,
             if self.expect(info_dict is not None, "Unable to retrieve transceiver {} info".format(i)):
                 if self.expect(isinstance(info_dict, dict), "Transceiver {} info appears incorrect".format(i)):
                     actual_keys = info_dict.keys()
-
```
Comment (mihirpat1) [inline tests/platform_tests/api/test_sfp.py:265]: Please revert this

Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:265]: I have reverted already and committed.

Comment (rajann) [reply tests/platform_tests/api/test_sfp.py:265]: This is the fix to address the addtional line created in previous commit

**Review by mihirpat1 — APPROVED** [review]

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@yxieca please help cherry pick to 202205

**Comment by prgeor** [comment]
@StormLiangMS please pick this change to 202211. Thanks

**Pull request merged**